Noise models

sophiaponte opened the pull request:
#175 

**channel-qvm.lisp** is an implementation of a qvm that requires the user to explicitly define a noise model. The noise model is a set of noise rules. The noise rules describe where different types of noise should be injected in a program. Each rule is made up of a noise-predicate (describes the 'where' part) and operation-elements (kraus operators describing the noise). During each `transition`, the channel-qvm checks to see if the current instruction matches any of the `noise-rules`. If there is a match, the channel-qvm applies the `noise-rule`'s operation elems either before or after the instruction is applied. 

```
(defun build-noise-model ()
  (let* ((pred1 (make-noise-predicate (match-strict-qubits (list 0 1)) 1 "after"))
          (pred2 (make-noise-predicate (match-strict-gates (list "X" "Z")) 1 "after"))
          (kraus-elems1 (list (generate-damping-kraus-ops 2 1)))
          (kraus-elems2 (list (generate-damping-dephasing-kraus-ops 3 1)))
          (rule1 (make-noise-rule pred1 kraus-elems1))
          (rule2 (make-noise-rule pred2 kraus-elems2)))
    (make-noise-model (list rule1 rule2))))
```

**approx-qvm.lisp** Is an implementation of a qubit level noise model qvm. The noise model for this qvm is specified by adding T1, T2 and fReadout values for each qubit. The approx-qvm will generate the kraus operators for these values (or assignment probabilities in the case of fRo) and save them. Then, after each gate application instruction, this qvm searches for kraus operators belonging to the current qubit for either T1 or T2 and will apply them. 

```
(defun make-noisy-approx-qvm ()
  (let* ((qvm (make-instance 'approx-qvm :number-of-qubits 2 :avg-gate-time 1)))
    (set-qubit-t1 qvm 0 5)
    (set-qubit-t2 qvm 0 4)
    (set-qubit-t1 qvm 1 2)
    (set-qubit-t2 qvm 1 3)))
```

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+    :initarg :kraus-ops
+    :accessor kraus-ops
+    :initform (make-hash-table :test 'eql)
+    :documentation "list of lists of kraus operators : ((t1 kraus ops) (t2 kraus ops) ...)"
```
Doc says `list`, `:init-form` says `hash-table`.

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+   and adds the kraus operators representing that noise to the kraus-ops slot. "
+  (declare (ignore args)) 
+  (format t "approx-qvm~%")
+  (if (t1 qvm) 
```
If you `if` has no else branch, a `when` is clearer.

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+(defclass approx-qvm (kraus-qvm)
+  ((t1
+    :initarg :t1
+    :accessor t1
```
Prefix accessors with the class name. e.g. `approx-qvm-t1`

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+(defmethod apply-all-kraus-ops ((qvm approx-qvm) (instr quil:gate-application))
+  "Apply all the kraus operators (for all of the noise sources) to the state of the program"
+  
+  (loop for kops being each hash-value of (kraus-ops qvm)
```
A nitty nit pick. We follow the style of prefixing `loop` keywords with a `:`. e.g. `(loop :for i :below 10 :collect i)`

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+that instruction. 
+"
+  (let ((gate   (pull-teeth-to-get-a-gate instr))
+        (params (mapcar #'(lambda (p) (force-parameter p qvm))
```
You don't need the `#'` here.

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+           (prob-k1 (/ prob 2))
+           (prob-k0 (- 1 prob-k1 ))
+           (kraus-ops (loop :for mat :in '("I" "Z")
+                 :for pj :in (list prob-k0 prob-k1)
```
Looks like this needs indenting.

notmgsk commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+to the t1 noise. "
+  
+  (let* ((ones 0)
+        (qubit 0)
```
off-by-one (space)

notmgsk commented on src/kraus-qvm-2.lisp:
```diff
@@ -0,0 +1,299 @@
+
+
+(defun match-strict-qubits (qubits)
+  #' (lambda (i-qubits i-name) (equal qubits i-qubits)))
```
delete spaces after `'`

notmgsk replied:
actually just delete `#'` since it's not required here

notmgsk commented on src/kraus-qvm-2.lisp:
```diff
@@ -0,0 +1,299 @@
+
+
+(defmethod instr-intersect (l1 l2)
+  (loop named outer :for inst1 in l1
```
colons for keywords

notmgsk commented on src/toy-qvm.lisp:
```diff
@@ -0,0 +1,158 @@
+      :do (incf results-i (do-noisy-measurement qvm q program1))
+      :do (incf results-x (do-noisy-measurement qvm q program2))
+          )
+    (let ((p00 (- 1 (float (/ results-i numshots))))
```
If you define `results-i` as a float (either with `0.0` or `0d0` depending on your need) then you can get rid of the explicit casting.

kilimanjaro commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,173 @@
+
+
+(defun fro-map (qubit fro)
+  (setf m (make-hash-table))
```
Just a warning, Common Lisp has several notions of object equality, and hash tables allow for this choice with the `:test` keyword argument. The default is `'eql`, which is fine for integers. If you are keying `quil:qubit` objects this won't work FYI

sophiaponte replied:
For now the qubit here is an int, but i'll keep this in mind if that changes

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+    :initarg :noise-predicate
+    :accessor noise-predicate
+    :initform nil 
+    :documentation "a function that is true when an instruciton is matched, false otherwise")
```
```suggestion
    :documentation "A function that is true when an instruction is matched, false otherwise")
```

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+    :initarg :operation-elements
+    :accessor operation-elements
+    :initform ()  
+    :documentation "operation elements (kraus operators) to apply when the predicate is satisfied")))
```
```suggestion
    :documentation "Operation elements (kraus operators) to apply when the predicate is satisfied")))
```

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+(defun make-noise-rule (predicate operation-elements)
+  "Returns a noise rule with the specified prediate and operation elements. "
+  (make-instance 'noise-rule :noise-predicate predicate :operation-elements operation-elements))
```
Slight readability preference for having this span several lines

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+(defmethod instr-intersect (l1 l2)
+  "Check if there are any gates in common between l1 and l2. 1 and l2 are  lists of string."
+  (loop named outer :for inst1 in l1
```
Use keyword symbols for loop controls (e.g. `:do`)

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+
+(defmethod instr-intersect (l1 l2)
+  "Check if there are any gates in common between l1 and l2. 1 and l2 are  lists of string."
```
Prefer capitalization of symbol names in docstrings, e.g. "L1" and "L2"

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+
+;; The following are examples of predicates! 
+(defun match-strict-qubits (qubits)
```
Depending on how often and in which manner this will get used, you might have a preference for `(&rest qubits)` here. This would allow you to write `(match-strict-qubits 0 1)` rather than `(match-strict-qubits (list 0 1))`. If you did want to apply _that_ to a list of qubits, then you could `(apply #'match-strict-qubits (list 0 1))`

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+;; The following are examples of predicates! 
+(defun match-strict-qubits (qubits)
+  "The returned function is true if the instruciton's qubits are exactly equal to qubits"
+  (lambda (i-qubits i-name) (equal qubits i-qubits)))
```
When we wrote the pyquil version of this we could be a bit sloppy because quil in pyquil has less of a clear AST. But here we really should have predicates working on `quil:instruction` objects (see class definitions in `quilc/src/ast.lisp`). For example, this would be something like

```
(lambda (instr)
  (and (typep instr 'quil:gate-application)
           (equal qubits (mapcar #'quil:qubit-index (gate-application-arguments instr)))))
```
or something like that
```

sophiaponte replied:
Yeah this sounds good. It's definitely cleaner and the way I was doing it before gave me "unused argument" compilation warnings if I didn't use both lambda arguments in the lambda statement.

appleby replied:
Just for future reference, you can silence those unused variable warnings with a `(declare (ignore blah))`. In this instance, e.g.,

```lisp
(lambda (i-qubits i-name) 
  (declare (ignore i-name))
  (equal qubits i-qubits))
```

kilimanjaro commented:
Overall this is good progress. Most of the things I mentioned are just minor stylistic tweaks to make this look more "idiomatic". The only more substantial issue is the predicate stuff. Here I think predicates should be basically any boolean function of a `quil:instruction`. This will take a bit more writing to get working at once, but is more powerful and more sustainable than the previous pyquil approach.

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+  "Check if there are any gates in common between l1 and l2. 1 and l2 are  lists of string."
+  (loop named outer :for inst1 in l1
+        do (loop named inner :for inst2 in l2
+                 do (when (string= inst1 inst2) (return-from outer "A")))))
```
Would `(intersection l1 l2 :test #'string=)` work for you here instead of these nested `loop`s?

sophiaponte replied:
yup! I didn't know about using `:test`, and intersection wasn't working because it wasn't testing for string equality. Thanks!

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+
+(defmethod initialize-instance :after ((qvm channel-qvm) &rest args)
+  " Initializes an instance of a krausqvm"
```
```suggestion
  "Initialize an instance of a CHANNEL-QVM"
```

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+  ;; For compiled measurements, refer to the source of that
+  ;; instruction.
+  (:method ((qvm channel-qvm) (instr compiled-measurement))
+    (apply-classical-readout-noise qvm (source-instruction instr))))
```
I think you can replace this `defgeneric` with a series of `(defmethod apply-classical-readout-noise ...)`, unless there is some reason you want to keep `apply-classical-readout-noise` and `apply-classical-readout-noisee` separate for debugging or whatever for now.

sophiaponte replied:
nope, that was just a typo I forgot to fix :(

sophiaponte replied:
I copied this function to match how readout noise is applied in `noisy-qvm.lisp`

appleby replied:
Makes sense. If it's just a typo and you don't want a separately-named generic function, then you probably want to split out all the `(:method ...)` forms here into their own toplevel `(defmethod ...)` forms (except for the one specialized on `pure-state-qvm` which you can drop since it is already defined in `noisy-qvm.lisp`). While you're at it, you might want to move the methods that are specialized on `approx-qvm` into the file where that class is defined. Check out the `(defmethod apply-classical-readout-noise ...)` in density-qvm.lisp for an example.

The syntax above is a convenient shorthand that lets you define a generic function and bunch of related methods all in one go, rather having a separate `defgeneric` followed by a bunch of `defmethod`s.  But if the generic function is already defined (in noisy-qvm.lisp), then this will replace its existing methods (assuming you fix the typo and change the name to match the one in noisy-qvm.lisp), which probably isn't what you want.

sophiaponte replied:
Ok, sounds good. In case we end up deprecating the defgeneric in `noisy-qvm.lisp`, I'll also specify a defmethod for the `pure-state-qvm` option and move it into `qvm.lisp`, so that I don't accidentally forget to handle the `pure-state-qvm` case

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-operators)
```
Not strictly required, but I *think* (could be wrong) we follow the convention of always explicitly specifying an explicit `defgeneric` for any methods that we add. Search, e.g. `defgeneric run` in `src/qam.lisp`.

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+
+
+;; TODO declare inlinabldee, turn on optimization
+(defun perturb-measurement (actual-outcome p00 p01 p10 p11)
```
Duplicate of a function in `noisy-qvm.lisp`. Ditto `check-povm` below. Don't forget to delete these before merging.

appleby commented on qvm.asd:
```diff
@@ -79,4 +79,6 @@
-               (:file "misc")))
+               (:file "misc")
+               (:file "channel-qvm")
+               (:file "toy-qvm")))
```
Looks like `toy-qvm` should be changed to `approximate-qvm.lisp`?

sophiaponte replied:
yup, the `toy-qvm` was an initial lisp exercise I did with Robert.

sophiaponte commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+  "Logical AND of 2 noise predicates"
+  (let* ((conjunction (alexandria:conjoin (predicate np1) (predicate np2)))
+         (priority (max (priority np1) (priority np2)))
+         (noise-position (noise-position np1)))
```
Yeah, thats a good point. I was debating a way to allow the user to chose which predicate's noise position would be applied, but maybe for now I'll just keep it the way it is and add a comment.

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,173 @@
+(defun simple-test-approx ()
+  (let* ((ones 0)
+         (qvm (make-instance 'approx-qvm :number-of-qubits 2 :avg-gate-time 1))
+         (numshots 100)
```
forgot to delete, whoops

sophiaponte commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+                                           (predicate-and (noise-predicate rule1) (predicate-not (noise-predicate rule2))) 
+                                           (operation-elements rule1))
+                                 :collect (make-noise-rule 
+                                           (predicate-and (predicate-not (noise-predicate rule1)) (noise-predicate rule1)) 
```
nice catch!

ecpeterson commented:
I gave this a once-over, and it looks like it's headed very much in the right direction. Please ping me to check back again after you've knocked out Erik's and Appleby's comments.

Exciting changes ahead for noise in the QVM!

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,374 @@
+    (apply-classical-readout-noise qvm (source-instruction instr))))
+
+
+(defun %corrupt-readout (qvm instr povm-map)
```
IIUC, it looks like this doing the same thing as `%corrupt-qvm-memory-with-povm` in noisy-qvm.lisp, except that it takes the `povm-map` as an argument rather than expecting the `qvm` to have a `readout-povms` method?

My $0.02: I think it would be OK to modify the definition of `%corrupt-qvm-memory-with-povm` in noisy-qvm.lisp to take the `povm-map` as an argument and then just call it rather than having this separate function.

Alternatively, I think you could define a `readout-povms` method for the `channel-qvm` like so (warning untested):

```lisp
(defmethod readout-povms ((qvm channel-qvm))
  (readout-povms (noise-model qvm)))
```

then use `%corrupt-qvm-memory-with-povm` as-is and get rid of `%corrupt-readout`.

sophiaponte replied:
Yup -- `%corrupt-readout` in `noisy-qvm.lisp` takes `povm-map` as a param whereas `%corrupt-qvm-memory-with-povm` expects the qvm to have a `readout-povms` method.  

I chose to "copy" the noisy-qvm.lisp function here because if these noise models eventually replace noisy-qvm, then I thought it would be better to have this separate function in the `channel-qvm` file. If it would be cleaner for me to just target `corrupt-qvm-memory-with-povm` then I can go ahead and change this.

appleby replied:
That's a good point. My preference would be for a single function if it can be made to work without too much hassle, but either is fine I guess.

sophiaponte replied:
Ok, I think a good solution is to just change the `noisy-qvm.lisp` `apply-classical-readout-noise` to just target %corrupt-readout instead of %corrupt-qvm-memory-with-povm. That way if we end up deprecating `noisy-qvm` we don't have to make changes to `channel-qvm.lisp`

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,173 @@
+  (loop :for source in (list (t1-ops qvm) (t2-ops qvm))
+        :do (cond ((= (list-length qubits) 1) 
+                   (let ((kraus-map (gethash (nth 0 qubits) source)))
+                     (when kraus-map (progn (check-kraus-ops kraus-map)
```
yes definitely

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,173 @@
+(defun generate-damping-dephasing-kraus-ops (t2 gate-time)  
+  "Calculates the kraus operators for dephasing and damping noise from t2, which
+is decoherence time. "
+  (generate-dephasing-kraus-ops t2 gate-time)) 
```
Yeah -- I am pretty sure the only difference between `damping-dephasing` and `dephasing` is that pure `dephasing` uses `tphi` and `damping-dephasing` uses t2. I made two separate methods because I wanted to make absolutely sure that the calculations are the same before combining them into the same function. I'll add a clarifying comment for `damping-dephasing` while I do some more research to figure that out

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,336 @@
+          :return (operation-elements rule)))
+
+
+(defgeneric apply-all-kraus-operators (qvm instr kraus-operators)
```
It seems slightly odd that the generic function's lambda list here is `(qvm instr kraus-operators)`, but the method specialized on `approx-qvm` has lambda list (ignoring specializers) `(qvm instr qubits)`.

One pattern I've seen is to have the generic function accept a final `&rest args`, then each method can unpack that to whatever it's expecting: qubits for `approx-qvm` or `kraus-operators` for `channel-qvm`. Alternatively, you can get a similar effect with `&key qubits kraus-operators`, etc.

Or maybe it's normal to have methods that expect different things from their required args, and I'm just not used to seeing it. ping @stylewarning for his thoughts.

sophiaponte replied:
I was able to change the lambda list for `approx-qvm` to be `(qvm instr kraus-ops)` instead of `(qvm instr qubits)` so that it matches the lambda list for the generic function and the `channel-qvm` method.

appleby commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,30 @@
+
+
+(defun make-test-channel-qvm ()
+  (let* ((p1 (make-noise-pred (match-any-n-qubits 1 0 1) 1 "after"))
```
I'm not opposed to the switch to `&rest` arguments for these `match-*` functions in general, but in this case I think it makes it harder to distinguish between the `n` argument and the `&rest qubits` args.

sophiaponte replied:
Yeah, I was thinking that too. I think maybe I should `&rest` the other `match-*` predicates, and just not this one.

ecpeterson commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,336 @@
+  (format t "applying ~a~%" instr)
+  (let* ((amps (%trial-amplitudes qvm))
+         (r (random 1.0d0))
+         (summed-p 0.0d0)
```
It's strange to name a non-boolean `-p`.

ecpeterson commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,336 @@
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj) amps (apply #'nat-tuple qubits))
```
This feels like an expensive loop. Can you just `incf summed-p`, then `:until .. :return kj`, then apply `kj`?

sophiaponte replied:
I copied this function from the previous implementation in `noisy-qvm.lisp`. When reading the comments in that file (which I copied into this one), I think you need to apply `kj` in each iteration of the loop

sophiaponte commented on qvm.asd:
```diff
@@ -79,4 +79,6 @@
-               (:file "misc")))
+               (:file "misc")
+               (:file "channel-qvm")
+               (:file "approximate-qvm"))
```
thanks for catching that :)

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,124 @@
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+; For now just use a single gate time value for all calculations
```
I will definitely add a more detailed comment here. Basically, the gate time is a parameter used in calculating the kraus operators for a T1 or T2 decay. Ideally, this gate-time value would be the gate time of the instruction preceding the application of the T1 or T2 kraus operators. Since I didn't know how to get the gate time for an qvm gate, for now, I decided to just let the gate time be a settable field in the approx-qvm.

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,124 @@
+
+
+(defmethod apply-all-kraus-operators ((qvm approx-qvm) (instr quil:gate-application) qubits)
+  "Apply all the kraus operators (for all of the noise sources) to the state of the program. If we have just encountered a 2-qubit gate, we need to tensor the kraus operators for each qubit's error source. Currently, this does not support > 2 qubit gates. Note: if the current instr is a 2 qubit gate and there is only kraus ops for one of the qubits, then no error is applied. For example, if q0 ha ...
```
Im sorry, not sure what you mean by this. If there are no error channels provided, then `(list (t1-ops) (t2-ops))` will be nil, and there will be nothing to loop over, so `apply-kraus-ops` never gets called.

ecpeterson replied:
I mean the `Note:` part. If one of the `gethash`es returns `nil`, I guess the behavior of `kraus-kron` is also to return `nil`. Why shouldn't it kron against the identity when it's passed a `nil` instead?

sophiaponte replied:
Gotcha, that makes a lot more sense

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,124 @@
+                        (quil:application-parameters instr)))
+        (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr))))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-operators qvm instr qubits)
```
The generic method for transition is just a pass-through. Not sure how to answer your last question :(. I thought that specifying instr is a `quil:gate-application` would be the right thing to do because I only want to apply gates (and potentially kraus operators) after gate application instructions.

ecpeterson replied:
This is a fine answer: I was worried that there _wasn't_ a fully generic definition for this function, in which case passing a MEASURE would throw an error. If there is a generic definition, I feel all set.

sophiaponte commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,336 @@
+
+
+(defmethod add-noise-models (nm1 nm2)
+  " Combines two noise models by concatenating their rules. This way, either rule can be matched by an instruction. "
```
The noise rules of the two noise models should be in `priority` order. I haven't implemented functionality for applying rules by priority (which is why this is currently wrong) -- apologies for not making that clear.

appleby commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,128 @@
+;; calculations are the same.
+(defun generate-damping-dephasing-kraus-ops (t2 gate-time)  
+  "Calculates the kraus operators for dephasing and damping noise from T2, which is decoherence time. "
+  (generate-dephasing-kraus-ops t2 gate-time)) 
```
One thing that's tickling my funny bone about this: here we are passing T2 through unmodified to `generate-dephasing-kraus-ops`, which claims it wants a `t-phi` as it's first arg. Are T2 and T-PHI the same thing? Or interchangeable in this case? The `tphi` function above suggests that T-PHI is some combination of T1 and T2. I'm sure this ties in to your comment above about double checking to make sure the calculations are equivalent. I just wanted to point out the t2 vs tphi thing.

sophiaponte replied:
From what I understand, T-PHI is pure dephasing time for a qubit, and T2 is dephasing time with T1 (amplitude damping time) wrapped in. So both T2 and T-PHI are measures of dephasing time, but they are not equivalent. I think the way to generate the kraus operators for T2 dephasing vs T-PHI dephasing is the same calculation, but I am not sure, which is what my comment refers to above.

appleby replied:
Fair enough. To be clear I have no real understanding of these calculations and therefore no good reason to think they aren't equivalent; I was just pattern matching on the documented "types" of the arguments.

sophiaponte commented on tests/approximate-qvm-tests.lisp:
```diff
@@ -0,0 +1,78 @@
+          :do (qvm::bring-to-zero-state (qvm::amplitudes approx-qvm))
+          :do (run approx-qvm)
+          :do (incf ones (qvm::dereference-mref approx-qvm (quil:mref "R0" qubit))))
+    (is (and (< ones numshots) (> ones 0)))))
```
Yeah -- and I am not sure why. It works fine when I run it on its own but as soon as I run it with the rest of the tests it fails! Maybe its modifying some underlying state or something...Idk, need to investigate more.

sophiaponte replied:
fixed it! Somewhere, the `*compile-before-running*` flag was being set which caused the tests to fail.

appleby replied:
oof, nice catch.

sophiaponte commented on tests/approximate-qvm-tests.lisp:
```diff
@@ -0,0 +1,78 @@
+      (is (equalp t1-q0 (gethash 0 (qvm::t1-ops approx-qvm))))
+      (is (equalp t1-q1 (gethash 1 (qvm::t1-ops approx-qvm))))
+      (is (equalp t2-q0 (gethash 0 (qvm::t2-ops approx-qvm))))
+      (is (equalp t2-q1 (gethash 1 (qvm::t2-ops approx-qvm)))))))
```
In this case the elems of the matrices should be exactly identical bc the same function is being used to generate them.

appleby replied:
That makes sense. I still feel like `equalp` is depending on internal details of the `magicl:matrix` representation though, and that an explicit elementwise test that only depends on the exported magicl interface would be better.  For instance, I think these `equalp` tests will likely break after Cole's "Magicl NG" PR gets merged, since `magicl:matrix` is switching from a `defstruct` to a `defclass`. That's my $0.02 anyway.

sophiaponte commented on tests/approximate-qvm-tests.lisp:
```diff
@@ -0,0 +1,78 @@
+    (is (= (length kron) 4)))
+   (let* ((k (qvm::generate-damping-kraus-ops 2 5))
+          (kron-first (qvm::kraus-kron k nil))
+          (kron-second (qvm::kraus-kron nil k)))
```
should return `nil`, which it does

appleby replied:
sorry, my comment was unclear. maybe add a test for that case as well?

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,128 @@
+  (%corrupt-readout qvm instr (readout-povms qvm)))
+
+
+(defun tphi (t1 t2)
```
I meant to have this as a helper function in case a user wanted to calculate `t-phi` from t1 and t2.

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,417 @@
+  ;; Apply all the kraus operators to the state of the system.
+  (loop :for kraus-map :in kraus-maps
+        :do
+           (check-kraus-ops kraus-map)
```
Would it suffice to move this `check-kraus-ops` into `make-noise-rule` ? And maybe change the `:accessor` for the `operation-elements` slot to a `:reader` ? In general I'm all for safety checks, but would be good if it can be safely moved out of this loop.

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,417 @@
+           (apply-kraus-ops qvm instr kraus-map)))
+
+
+(defmethod apply-kraus-ops ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
```
Does this need to be a generic function?  It seems to only be called from within `apply-all-kraus-operators`, which has already taken care of specialization on qvm and instruction type.  If possible, consider making it a regular function and renaming to make it clear it's a "helper" for `apply-all-kraus-operators`, like `%apply-one-kraus-operator` or some such.

sophiaponte replied:
I wanted to have it specialized originally because I am anticipating incorporating different state-representations into the noisy qvms. When this happens, `apply-kraus-ops` will have to specialize on the type of state-representation, as applying kraus ops to a density matrix will be different than applying kraus ops to a pure state. If you think it would be better for me to make this a regular function now and just deal with changing it to a method in a future PR when I actually need to, I can definitely do that.

appleby replied:
That makes sense. I think it's fine to leave as a generic function if you're pretty sure you'll need it to be in the near future.

sophiaponte commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,206 @@
+         (nm (qvm::make-noise-model (list nr1 nr2))))
+    (setf (qvm::qubit-povm nm 0) (list .9d0 .1d0 .1d0 .9d0))
+    (setf (qvm::qubit-povm nm 1) (list .8d0 .2d0 .2d0 .8d0))
+                                        ; check that rules are sorted in priority order
```
I'm confused what you mean here

appleby replied:
Sorry. I meant add another semi-colon to the start of the comment, like `;; check that...`.

The (admittedly somewhat arcane) general rule for comment characters is:

- 4 semi-colons to precede file-header comments and sometimes "major section" headers
- 3 semi-colons for top-level comments that aren't a file header
- 2 semi-colons for comments indented with code
- 1 semi-colon for end-of-line comments

Believe it or not, these comment conventions are even enshrined the spec! Almost brings a tear to the eye it's so beautiful...

http://www.lispworks.com/documentation/HyperSpec/Body/02_ddb.htm

Your editor knows about these conventions and therefore indented the single-semi-colon comment way over to the right thinking you wanted an end-of-line comment. This is `comment-column` in emacs.

sophiaponte commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,417 @@
+
+(defmethod make-noise-pred (predicate priority noise-position)
+  ;; Creates a NOISE-PRED with the given PREDICATE (callable),
+  ;; PRIORITY, and NOISE-POSITION
```
I thought methods had to have comments and funcs had to have docstrings? Should I be making this a function instead of a method?

appleby replied:
Oops! You are correct. Typically, the docstring for a method gets attached to the corresponding generic function.  I actually didn't notice this was a method. Without searching through the diff to see how it's used, it should probably be a regular function if this is the only method you've defined. As a general rule, `make-foo` style functions are usually regular functions.

Same goes for the `predicate-blah` methods below. I think those can all be regular functions?

Generally speaking, use generic functions and methods if you want to dispatch on the type (or in some limited cases the value) of one or more arguments, or if you're overriding some pre-existing method.

appleby commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,417 @@
+
+(defmethod add-noise-models (nm1 nm2)
+  ;; Combines two noise models by concatenating their rules. This way,
+  ;; either rule can be matched by an instruction.
```
Same thing here re: `defmethod` vs `defun`, assuming there are no other methods defined.

sophiaponte commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,417 @@
+   (slot-value qvm 'original-amplitude-pointer) (amplitudes qvm)))
+
+
+(defun requires-swapping-p (qvm)
```
Sounds good. I'll delete the duplicates in `noisy-qvm.lisp`

sophiaponte commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,132 @@
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q in qubits :collect (gethash q (t1-ops qvm))))
+         (t2-ops (loop :for q in qubits :collect (gethash q (t2-ops qvm))))
+         (kraus-ops (list t1-ops t2-ops)))
```
Sure, I could do that! Would that be better to do stylistically? I originally thought it would be better for each `apply-all-kraus-operators` to be specialized on each qvm class for readability. I thought each class should have its own method to clearly express "This is how the kraus operators are applied for each type of noisy qvm". 

In this case, since the `apply-all-kraus-operators` on `approx-qvm` is so short, is it better to just have that code exist in `transition`?

appleby replied:
Hmmm good question. I think having a separate function for applying the kraus ops is reasonable. I guess I just noticed that the difference between the methods for `channel-qvm` and `approx-qvm` is small now after the latest refactoring and wondered if they could be combined. It also struck me that the difference that remains isn't really in how the operators are *applied* so much as how they are extracted/combined/whatever, but that might just be arguing semantics. At the end of the day, it's stylistic question as you say, so whatever you think is the right way to slice it is fine by me.

I definitely think it makes sense to have something like the existing `apply-kraus-ops` since that needs to happen in several places and is non-trivial. In fact, we should probably be calling that from inside `noisy-qvm`'s `transition` method as well (assuming the implementations are identical, I didn't look too closely).

appleby replied:
Oh, I just realized after reading your other comment about `apply-kraus-ops` that `noisy-qvm` and `channel-qvm` are "siblings", and so calling `apply-kraus-ops` from `noisy-qvm`'s `transition` method probably wouldn't work (since it's specialized on `channel-qvm`).

sophiaponte commented:
Recent updates:
- I put the `NOISE-MODEL`, `NOISE-PREDICATE`, and `NOISE-RULE` classes and related functions into a separate file (noise-models.lisp). I also moved related tests into `tests/noise-model-tests.lisp`

- I added DEPOLARIZING-NOISE to APPROX-QVM, as it is a frequently used to model QPU noise.

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q :in qubits :collect (gethash q (t1-ops qvm))))
+         (t2-ops (loop :for q :in qubits :collect (gethash q (t2-ops qvm))))
+         (depol-ops (loop :for q in qubits :collect (gethash q (depol-ops qvm))))
```
`:in`

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (t1-ops (loop :for q :in qubits :collect (gethash q (t1-ops qvm))))
```
What if any of these lists contain `nil`?

If they never should, I think we should add `assert`s.

sophiaponte replied:
If one of the lists is NIL (ie, the user doesn't specify a T2 for a qubit in `instr`), then in `apply-all-kraus-maps` the NIL is just passed over (so nothing is done to the state, which is what should happen)

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If both T1 and T2 are 0, TPHI = 0."
+  (if (and (> t1 0) (> t2 0))   
+      (/ (* 2 (* t1 t2)) (+ t2 (* 2 t1)))
```
put denominator on a new line

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If both T1 and T2 are 0, TPHI = 0."
+  (if (and (> t1 0) (> t2 0))   
```
`(and (plusp t1) (plusp t2))`

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+  (let*  ((prob (/ gate-time t-phi))
+          (prob-k0 (/ prob 2))
+          (prob-k1 (- 1 prob-k0 ))
+          (kraus-ops (loop :for mat :in '("I" "Z")
```
no need for this variable

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+  "Given a value for T-PHI (dephasing time) and a GATE-TIME, calculates the kraus operators corresponding to the dephasing noise."
+  (let*  ((prob (/ gate-time t-phi))
+          (prob-k0 (/ prob 2))
+          (prob-k1 (- 1 prob-k0 ))
```
extra space at the end

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+          (kraus-ops (loop :for mat :in '("I" "Z")
+                           :for pj :in (list prob-k0 prob-k1)
+                           :collect (magicl:scale (sqrt pj)
+                                                  (quil:gate-matrix
```
I would factor this out into its own function.

stylewarning commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,183 @@
+  (assert (< 0 prob 1) (prob) "DEPOL-PROB must be between 0 and 1")
+  (let* ((pk0 (sqrt (- 1 (/ (* 3 prob) 4))))
+         (pkn (sqrt (/ prob 4)))
+         (kraus-ops (loop :for mat :in '("I" "X" "Y" "Z")
```
no need for this variable

stylewarning commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,209 @@
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-matched-rule (match-rules noise-rules instr :before))
```
use a quote on `':before` and `':after`

stylewarning commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,209 @@
+
+(defun match-rules (rules instr position)
+  "Return the first matching rule."
+  (loop :for rule :in rules
```
Use `find`

sophiaponte commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,263 @@
+  ;; Assert that the NOISE-POSITION of the predicate is of type
+  ;; NOISE-POS, and that PRIORITY is of type INTEGER
+  (declare (ignore args))
+  (assert (typep (noise-position pred) 'noise-pos))
```
I tried to change this, and got weird warnings. @stylewarning informed me that `check-type` requires the args to be SETF-able, which `noise-position` (and `priority` later down) are not.

appleby replied:
Oops, sorry! Yes, that's correct. I was thinking `noise-position` was a slot. Type arg should also not be quoted for `check-type` now that I see it again...

sophiaponte commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,180 @@
+
+
+(deftest test-match-rules ()
+  ;; Test that MATCH-RULES correctly returns the kraus operators for
```
yup, nice catch!!

sophiaponte commented on tests/approximate-qvm-tests.lisp:
```diff
@@ -0,0 +1,232 @@
+    (loop :for k in kron
+          :do (is (magicl::matrix-rows (nth 0 kron)) 4)
+          :do (is (magicl::matrix-cols (nth 0 kron)) 4))
+    (is (= (length kron) 4)))
```
After going through and changing a bunch, I definitely agree that it makes it more readable. Thanks for pointing it out to me :)

appleby replied:
🙌

sophiaponte commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,180 @@
+  ;; Test that noisy readout works on the channel qvm. Applying a
+  ;; program 100 times and evaluate that the resulting excited state
+  ;; population is affected by the reaodut noise.
+  (let* ((nm (make-instance 'qvm::noise-model))
```
The reason I added it in the approx-qvm tests was because somehow the program was being compiled before `run`. I think I solved the issue with the addition of some functions into `channel-qvm` from `noisy-qvm` to prevent programs from being compiled for channel/approx qvms and that seems to have worked. I think I'll add the guard in channel qvm as a safeguard as well.

stylewarning commented on tests/approximate-qvm-tests.lisp:
```diff
@@ -0,0 +1,207 @@
+(in-package #:qvm-tests)
```
add
```
;;;; filename.lisp
;;;;
;;;; Author: Sophia Ponte
```

as with the other files

stylewarning commented on src/qvm.lisp:
```diff
@@ -131,6 +131,9 @@ ALLOCATION is an optional argument with the following behavior.
   (bring-to-zero-state (amplitudes qvm))
   qvm)
 
+(defmethod apply-classical-readout-noise ((qvm pure-state-qvm) (instr quil:measurement))
```
isn't this method defined much later? When this DEFMETHOD is executed, it's not yet defined. I think it needs to move back to the DEFGENERIC.

stylewarning commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,262 @@
+                                       (operation-elements r2))))))
+
+
+(defun match-strict-qubits (&rest qubits)
```
I would rename these functions `match-...` to `...-matcher`. They don't literally match, but produce a boolean function that matches.

sophiaponte replied:
`...-predicate`

stylewarning commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,262 @@
+(alexandria:define-constant +DEFAULT-NOISE-PRED-NAME+ "NOISE-NAME" :test #'string=)
+
+
+(deftype noise-pos ()
```
prefer a fully spelled out name `noise-position`

stylewarning commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,262 @@
+  (:documentation "A NOISE-PRED describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PRED can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
+
+
+(defconstant +MAX-PRIORITY+ 20)
```
move these constants to the top of the file imho

sophiaponte replied:
also add comment

stylewarning commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,262 @@
+   (priority
+    :initarg :priority
+    :reader priority
+    :initform 10 ; Default priority is 10
```
do default-initargs for all

stylewarning commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,207 @@
+      ((1) (if (<= r p01) 0 1)))))
+
+
+;; Duplicate fn in noisy-qvm.lisp (deleted in noisy-qvm.lisp)
```
no need for "deleted in..." if there's only one place that this function lives

stylewarning commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,207 @@
+    (assert (cl-quil::double= 1.0d0 (+ p01 p11)))))
+
+
+(defun noisy-program-strings (parsed-program qvm)
```
im dubious of the value of returning a bunch of strings

sophiaponte replied:
do you think this would be better as a function that is exclusively used for testing? Like should I move this to a test file?

stylewarning replied:
yes

ecpeterson commented on src/approximate-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+;;; each qubit. During the APPROX-QVM's TRANSITION, qubit-level noise
+;;; is applied after each instruction for which qubit noise is
+;;; specified.
+(defclass approx-qvm (channel-qvm)
```
Not thrilled with this class name.

sophiaponte replied:
changed to `basic-noise-qvm`

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the T2  provided on the specified qubit."
```
```suggestion
  "Save the kraus operators for the T2 provided on the specified qubit."
```

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
```
```suggestion
  "Save the readout POVMs for the FRO provided on the specified QUBIT. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout POVM assignment probabilities."
```

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
```
This produces slightly better code: 
```suggestion
(let ((one-minus-fro (- 1.0d0 fro)))
  (setf (gethash qubit (readout-povms qvm)) (list fro one-minus-fro one-minus-fro fro)))
```
One would think that the compiler ought to notice that (- 1.0d0 fro) is constant and reuse the value but, alas, it computes it twice unless you introduce the `let` form.

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+
+(defmethod apply-all-kraus-maps ((qvm basic-noise-qvm) (instr quil:gate-application) kraus-ops)
+  ;; Apply all the kraus operators for each type of noise
```
This looks like a documentation string, rather than a comment.

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2). If both T1 and T2 are 0, TPHI = 0."
+  (if (and (plusp t1) (plusp t2))
```
Use appropriately-typed constants: 2.0d0 instead of 2 and 0.0d0 instead of 0. Same thing everywhere else in the code.

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+        ((not k2s) (loop :for k :in k1s
+                         :collect (magicl:kron k (magicl:make-identity-matrix 2))))
+        (t  (loop :for k1 :in k1s :append (loop :for k2 :in k2s
+                                                :collect (magicl:kron k1 k2))))))
```
Use NULL instead of NOT and compute the 2x2 identity just once:
```
  (cond ((null k1s) (loop :with identity-matrix := (magicl:make-identity-matrix 2)
                          :for k :in k2s :collect (magicl:kron identity-matrix k)))
        ((null k2s) (loop :with identity-matrix := (magicl:make-identity-matrix 2)
                          :for k :in k1s :collect (magicl:kron k identity-matrix)))
        (t (loop :for k1 :in k1s :append (loop :for k2 :in k2s
                                               :collect (magicl:kron k1 k2))))))
```

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+                (apply-kraus-map qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm basic-noise-qvm) (instr quil:gate-application))
```
Do you think it would make sense to use an `:after` method combination [1] here? This seems like a conventional transition on which you call `apply-all-kraus-maps` **after** you update the wavefunction. Using an `:after` combination would slightly simplify the code and, more importantly, convey the intention clearly.

[1] http://www.lispworks.com/documentation/HyperSpec/Body/07_ffb.htm

sophiaponte replied:
I just tried doing this because I agree that it makes sense. However, I ran into issues because `basic-noise-qvm` is a subclass of `channel-qvm`, which is a subclass of `pure-state-qvm`. If I make `basic-noise-qvm`'s transition an `:after` method, it tries to run `channel-qvm`'s transition first, which I do not want. Unless there is a way to prevent this from happening and to default all the way back to `pure-state-qvm`'s transition, then I think I have to keep it the way it is.

jmbr replied:
I see. Even though we could probably create a suitable method combination to accomplish this goal, for the moment I think this can be left as is (perhaps adding a short comment explaining that an `:after` method won't do in this case).

sophiaponte replied:
I will make an issue for refactoring the `transition` functions for `channel-qvm` and `basic-noise-qvm` to use `:around`/`:after` to avoid repetitive code, but for now I'll leave them as they are

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+;;; apply kraus operators to it's state.
+
+(defgeneric apply-all-kraus-maps (qvm instr kraus-maps)
+  (:documentation "Applies every kraus-map in the list KRAUS-MAPS to the state of the system."))
```
```suggestion
  (:documentation "Applies every kraus map in the list KRAUS-MAPS to the state of the system."))
```

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+
+
+(defgeneric apply-kraus-map (qvm instr kraus-ops)
+  (:documentation "Applies noise from a kraus map to the system. Randomly select a kraus operator from the kraus map KRAUS-OPS to apply to the state of the system."))
```
Specify what you mean by "randomly." My guess is you mean "uniformly at random."

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+;;; noise should be applied. These decisions are made during
+;;; TRANSITION. At each new instruction, the CHANNEL-QVM checks to see
+;;; if any NOISE-RULES are matched by the instuction, and if so, the
+;;; CHANNEL-QVM applied the corresponding noise.
```
```suggestion
;;; CHANNEL-QVM applies the corresponding noise.
```

jmbr replied:
Suggestion: blend the parts of this description that don't deal with the nitty-gritty into the documentation string.

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+   (noise-model
+    :initarg :noise-model
+    :accessor noise-model
+    :initform nil))
```
Suggestion: move this into `:default-initargs` for consistency with the rest of our code.

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+;;; if any NOISE-RULES are matched by the instuction, and if so, the
+;;; CHANNEL-QVM applied the corresponding noise.
+(defclass channel-qvm (pure-state-qvm)
+  ((original-amplitude-pointer
```
Maybe call this `original-amplitudes` instead? Calling it a pointer gives me the heebie jeebies :-)

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  ;; If any noise rules are matched by theinstruction INSTR, apply the
```
```suggestion
  ;; If any noise rules are matched by the instruction INSTR, apply the
```

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
```
As with `transition` in `basic-noise-qvm`, I wonder if this is the best method combination. Consider what this would look like using `:around`.

sophiaponte replied:
https://github.com/rigetti/qvm/pull/195#discussion_r342281033

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+
+(defun find-matching-rule (rules instr position)
+  "Return the first rule that matches INSTR/POSITION in the list RULES."
+  (find-if (lambda (rule) (rule-matches-instr-p rule instr position)) rules))
```
Is a list the most appropriate data structure for the rule set? This query is going to be O(n) where n is the size of the rules. Probably OK for this PR but we might need to revisit this in the future.

sophiaponte replied:
I was also thinking about this, but I agree that for at least this PR a list should be okay. For most use cases that I can think of, the list of noise rules will be small compared to the length of the program.

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+         (summed 0.0d0)
+         (logical-qubits (quil:application-arguments instr))
+         (qubits (mapcar #'quil:qubit-index logical-qubits)))
+    (check-type amps quantum-state)
```
The following change:
```suggestion
    (declare (type quantum-state amps)
```
will be equivalent to `check-type` (i.e., it will trigger an error) when the `safety` value of the current compiling policy is sufficiently high and will turn into a no-op (i.e., it will do nothing) when the `speed` value of the policy is sufficiently high and `safety` is low.  This function is called on each Kraus map twice per call to `transition`, so I think it is preferable to replace the (unconditional) `check-type` by a type annotation that is selectively enforced by the compiler.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+
+;;; Implementation of rule-based noise models for the QVM.
+
+;;; These values are relatively aribtrary limits for NOISE-PREDICATE
```
```suggestion
;;; These values are relatively arbitrary limits for NOISE-PREDICATE
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+;;; instructions.
+(defconstant +MAX-PRIORITY+ 20)
+(defconstant +MIN-PRIORITY+ 0)
+(alexandria:define-constant +DEFAULT-NOISE-PREDICATE-NAME+ "NOISE-NAME" :test #'string=)
```
No need to capitalize:
```suggestion
(alexandria:define-constant +default-noise-predicate-name+ "NOISE-NAME" :test #'string=)
```

sophiaponte replied:
should +MAX-PRIORTY+ and +MIN-PRIORITY+ also be lower case?

jmbr replied:
I'd vote for that!

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+   :priority 10
+   :noise-position ':after
+   :name +DEFAULT-NOISE-PREDICATE-NAME+)
+  (:documentation "A NOISE-PREDICATE describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATEICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PREDICATE can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
```
```suggestion
  (:documentation "A NOISE-PREDICATE describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PREDICATE can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  (:documentation "A NOISE-PREDICATE describes a collection of program instructions. When used in a NOISE-RULE, the NOISE-PREDICATEICATE describes which instructions should be matched by the noise described in the OPERATION-ELEMENTS. Optionally, the NOISE-PREDICATE can be given a NAME. If none is provided, use the default 'NOISE-NAME'."))
+
+
+(deftype noise-position ()
```
I'd move the declaration of this type so that it precedes the declaration of the `noise-predicate` class to help the (human) reader but it's no big deal.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  ;; NOISE-POSITION, and that PRIORITY is of type INTEGER
+  (declare (ignore args))
+  (assert (typep (noise-position pred) 'noise-position))
+  (assert (typep (priority pred) 'integer))
```
`check-type` instead?

appleby replied:
haha. fooled me as well! apparently `noise-position` and `priority` are regular functions not slot readers.

https://github.com/rigetti/qvm/pull/195#discussion_r339880679

appleby replied:
Ah, I misunderstood again. Just checked. They *are* slots, just read-only.

jmbr replied:
Using `check-type` seems to work just fine:
```
(let ((noise-pred (qvm::make-noise-predicate (qvm::match-strict-qubits 0) 1 :after)))
       (check-type (noise-position noise-pred) noise-position)
       (check-type (priority noise-pred) integer))
```
The above returns `nil` and if I change the types in the `check-type` forms to something else, it signals an error.

appleby replied:
yes but if you try to use the `store-value` restart it will complain that the setf function is undefined. also gives style-warnings about the same on compilation for me.

jmbr replied:
Sounds like a fair amount of bending-over-backwards for a restart that no one is going to use. Plus, there's `(check-type (slot-value noise-pred 'priority) integer)`.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  (assert (typep (noise-position pred) 'noise-position))
+  (assert (typep (priority pred) 'integer))
+  (let ((pr (priority pred)))
+    (assert (<= +MIN-PRIORITY+ pr +MAX-PRIORITY+) (pr)
```
Again, purely cosmetic changes:
```suggestion
    (assert (<= +min-priority+ pr +max-priority+) (pr)
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  (assert (typep (priority pred) 'integer))
+  (let ((pr (priority pred)))
+    (assert (<= +MIN-PRIORITY+ pr +MAX-PRIORITY+) (pr)
+            "PRIORITY of a NOISE-PREDICATE must be between MIN-PRIORITY and MAX-PRIORITY")))
```
In this context, "priority" is a regular word, not a symbol name:
```suggestion
            "Priority of a NOISE-PREDICATE must be between MIN-PRIORITY and MAX-PRIORITY")))
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  ((noise-predicate
+    :initarg :noise-predicate
+    :accessor noise-predicate
+    :documentation "A function that is true when an instruciton is matched, false otherwise.")
```
```suggestion
    :documentation "A function that is true when an instruction is matched, false otherwise.")
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+
+
+(defun rule-priority>= (r1 r2)
+  "Comparator function for the priority of 2 noise rules. Returns true if the priority of R1 is <= to the priority of R2."
```
```suggestion
  "Comparator function for the priority of 2 noise rules. Returns true if the priority of R1 is >= to the priority of R2."
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  ((noise-rules
+    :initarg :noise-rules
+    :accessor noise-rules
+    :initform nil
```
Move to `:default-initargs`.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+   (readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)
```
Ditto.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+
+
+(defun match-instr-idxs (parsed-prog &rest idxs)
+  "The returned function is true if the index of the INSTR in the program PARSED-PROG matches IDX."
```
```suggestion
  "The returned function is true if the index of the INSTR in the program PARSED-PROG matches IDXS."
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+    :reader predicate-function
+    :documentation "Does the instruction trigger the application of a channel?")
+   (priority
+    :initarg :priority
```
Suggestion: create a `noise-priority` type that encompasses integers from `+min-priority+` to `+max-priority+`.

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+  ;; NOISE-POSITION, and that PRIORITY is of type INTEGER
+  (declare (ignore args))
+  (assert (typep (noise-position pred) 'noise-position))
+  (assert (typep (priority pred) 'integer))
```
Following the suggestion above, replace integer by the more specific `noise-priority` type.

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+(in-package #:qvm)
+
+;;; Implements a QVM that can interpret a rule-based noise model and
+;;; apply kraus operators to it's state.
```
```suggestion
;;; apply kraus operators to its state.
```

kilimanjaro commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,204 @@
+
+(in-package #:qvm)
+
+;;; Implements a QVM that can interpret a rule-based noise model and
```
This is to varying degrees documented elsewhere, but for the sake of orienting the reader I would suggest giving the reader a brief high-level picture of what is going on here (or someplace else, perhaps at the top of `noise-models.lisp`). It doesn't have to be much, but could include a brief summary of:
- what do you mean by a rule-based noise model? (mention the mechanism, that noise is applied either before or after a Quil instruction, based on syntactic properties of that instruction)
- how are models represented? no real details needed, can just refer to the NOISE-MODEL and NOISE-PREDICATE classes
- maybe a mention about the priority system (e.g. only one rule is applied in any context, the highest priority of the matching rules)

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+(in-package #:qvm)
+
+;;; Implements a QVM that simulates T1, T2, READOUT,  and/or depolarizing noise
```
```suggestion
;;; Implements a QVM that simulates T1, T2, READOUT, and/or depolarizing noise
```

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+;;; after every instruction in a program.
+
+
+;;; The A-QVM supports a QUBIT-level noise model, intended to
```
```suggestion
;;; The BASIC-NOISE-QVM supports a qubit-level noise model, intended to
```

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+;;; The A-QVM supports a QUBIT-level noise model, intended to
+;;; approximate the qubit noise on a QPU. The APPROX-QVM allows the
+;;; specification of T1, T2, DEPOLARIZATION, and READOUT noise for
+;;; each qubit. During the APPROX-QVM's TRANSITION, qubit-level noise
```
```suggestion
;;; each qubit. During the BASIC-NOISE-QVM's TRANSITION, qubit-level noise
```

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+
+;;; The A-QVM supports a QUBIT-level noise model, intended to
+;;; approximate the qubit noise on a QPU. The APPROX-QVM allows the
```
```suggestion
;;; approximate the qubit noise on a QPU. The BASIC-NOISE-QVM allows the
```

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+
+;;; The A-QVM supports a QUBIT-level noise model, intended to
+;;; approximate the qubit noise on a QPU. The APPROX-QVM allows the
+;;; specification of T1, T2, DEPOLARIZATION, and READOUT noise for
```
are these capitalized because they're symbols somewhere? Doesn't look like it.

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+  ((t1-ops
+    :initarg :t1-ops
+    :accessor t1-ops
+    :documentation "T1 noise for each qubit to simulate")
```
Very important for the docstring to give a sense of what type is there

stylewarning commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,194 @@
+(defun generate-depolarizing-kraus-map (prob)
+  "Given a probability of depolarization PROB, calculate the kraus operation elements for the depolarizing channel."
+  (assert (< 0 prob 1) (prob) "DEPOL-PROB must be between 0 and 1")
+  (let* ((pk0 (sqrt (- 1 (/ (* 3 prob) 4))))
```
`(- 1 (* 3/4 prob))` will work

sophiaponte commented:
In this last review I mostly made changes to documentation and a few smaller stylistic things. However, I did have to make a pretty big change in `BASIC-NOISE-QVM` to more accurately generate the Kraus operators for damping and dephasing noise. I'll summarize the changes here: 
- Damping and depolarizing noise is time dependent. Therefore, it doesn't make sense to precompute the kraus operators for this noise before running a qvm program. Now, the operators for these noise sources are calculated at every `transition`, and the kraus operators are parametrized by the elapsed time since the start of the program.  
- I added an `elasped-time` slot to the basic-noise-qvm which is incremented by `average-gate-time` after every instruction, to address the above point. 
- Dephasing noise is calculated from t-phi. I added a slot to the qvm so that users can provide t-phi values for each qubit. If these values are not provided, then t-phi is calculated from T1 and T2, if these are available.

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,216 @@
+;;; match qvm instructions, and which instructions to match. The
+;;; OPERATION-ELEMENTS is the physical description of the channel,
+;;; represented as kraus operators. At each qvm instruction, the
+;;; OPERATION-ELEMENTS of a NOISE-RULE who's NOISE-PREDICATE matches
```
```suggestion
;;; OPERATION-ELEMENTS of a NOISE-RULE whose NOISE-PREDICATE matches
```

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,216 @@
+
+;;; The CHANNEL-QVM requires an explicitly defined NOISE-MODEL. At the
+;;; TRANSITION to each new instruction, the CHANNEL-QVM checks to see
+;;; if any NOISE-RULES are matched by the instuction, and if so, the
```
```suggestion
;;; if any NOISE-RULES are matched by the instruction, and if so, the
```

jmbr commented on src/channel-qvm.lisp:
```diff
@@ -0,0 +1,216 @@
+    :accessor noise-model))
+  (:default-initargs
+   :noise-model nil)
+  (:documentation "The CHANNEL-QVM is a QVM that supports a fully explicit NOISE-MODEL. The NOISE-MODEL is an explicit definition of where and how different channels should be applied to a program running in the CHANNEL-QVM. list of NOISE-RULES, where each NOISE-RULE describes a type of noise and at which points in a program's execution that type of noise should be applied."))
```
Like this instead?
```suggestion
  (:documentation "The CHANNEL-QVM is a QVM that supports a fully explicit NOISE-MODEL. The NOISE-MODEL is an explicit definition of where and how different channels should be applied to a program running in the CHANNEL-QVM."))
```

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+
+
+(defmethod predicate-not (np)
+ "Logical NOT of a NOISE-PREDICATE NP. The PRIORITY of the predicate is 'inverted', by taking the difference between it and +maxpriority+."
```
Microscopic nit: indent this documentation string by one additional space.

jmbr replied:
Another nit: rename `np` to `noise-predicate` below and rewrite the docstring as:
```suggestion
 "Logical NOT of a NOISE-PREDICATE. The PRIORITY of the predicate is 'inverted', by taking the difference between it and +maxpriority+."
```
The rationale for doing this is that when you write code that calls `predicate-not`, Emacs will tell you the expected argument names and `NOISE-PREDICATE` says everything you need to know whereas `NP` is more cryptic. Same thing above with the other predicates.

This is just a suggestion and something for you to consider. It's not a deal-breaker for merging the code as far as I am concerned.

jmbr commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,249 @@
+
+(defun damping-kraus-map (t1 elapsed-time)
+  "Given a value for T1 and a ELAPSED-TIME, generates the kraus operators corresponding to the amplitude damping noise."
+  (let* ((prob (- 1 (exp (/ (* -1 elapsed-time) t1))))
```
```suggestion
  (let* ((prob (- 1 (exp (/ (- elapsed-time) t1))))
```
Same elsewhere and in the test suite.

appleby commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,271 @@
+(defun match-instr-idxs (parsed-prog &rest idxs)
+  "The returned function is true if the index of the INSTR in the program PARSED-PROG matches IDXS."
+  (lambda (instr)
+    (member (position instr (cl-quil::parsed-program-executable-code parsed-prog) :test 'eq) idxs)))
```
as mentioned in slack, the default test is `eql` (I think), which would also work here, but `eq` should be a tad faster

notmgsk replied:
oooooooooooooooooooooooöooooöps

jmbr commented on src/noise-models.lisp:
```diff
@@ -0,0 +1,277 @@
+;;; instructions.
+(defconstant +maxpriority+ 20)
+(defconstant +minpriority+ 0)
+(deftype noise-priority () '(integer 0 20))
```
```suggestion
(deftype noise-priority () '(integer ,+minpriority+ ,+maxpriority+))
```

appleby commented on tests/basic-noise-qvm-tests.lisp:
```diff
@@ -0,0 +1,233 @@
+      (setf (qvm::qubit-t1 q qubit) 2)
+      (setf (qvm::qubit-t2 q qubit) 1.5)
+      (let ((ones-measured (qvm-tests::run-n-shot-program numshots q program)))
+        (print ones-measured)
```
debug print?

appleby commented on tests/basic-noise-qvm-tests.lisp:
```diff
@@ -0,0 +1,233 @@
+         (t2 16)
+         (tphi (qvm::tphi t1 t2)))
+    (is (= 9.6d0 tphi)))
+  (signals error (qvm::tphi 0 1))
```
maybe add a test for `t1` = `t2` = 0 ?

appleby commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,188 @@
+           (q (make-instance 'qvm::channel-qvm :number-of-qubits 1 :noise-model nm))
+           (povm (list .9d0 .1d0 .1d0 .9d0)))
+      (setf (gethash qubit (qvm::readout-povms (qvm::noise-model q))) povm)
+      (let ((ones-measured (qvm-tests::run-n-shot-program numshots q program)))
```
```suggestion
      (let ((ones-measured (run-n-shot-program numshots q program)))
```

likewise below

appleby commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,188 @@
+      ;; 50. The higher priority rule, which is the one that we want to
+      ;; apply, should make ones > 50
+      (let ((ones-measured (qvm-tests::run-n-shot-program numshots q program)))
+        (print ones-measured)
```
debug print?

appleby commented on tests/channel-qvm-tests.lisp:
```diff
@@ -0,0 +1,188 @@
+            :collect (qvm::name (qvm::noise-predicate rule-after-instr)))))
+
+(deftest test-print-noisy-program ()
+  ;; This function tests that QVM::PRINT-NOISY-PROGRAM correctly
```
print-noisy-program -> noisy-program-strings

sophiaponte commented on src/basic-noise-qvm.lisp:
```diff
@@ -0,0 +1,256 @@
+  (check-type t2 valid-noise-value)
+  (if (or (= 0.0d0 t1) (= 0.0d0 t2)) 0.0d0)
+  (/ (* 2.0d0 (* t1 t2))
+     (+ t2 (* 2.0d0 t1))))
```
For some reason when I do 
```  (if (or (= 0.0d0 t1) (= 0.0d0 t2))
      0.0d0
      (/ (* 2.0d0 (* t1 t2))
         (+ t2 (* 2.0d0 t1)))))```
I get a compiler warning "deleting unreachable code". So I did this instead. It seems to work -- if the `(or ...)` fails, it returns the division computation.

rigettizach replied:
The unreachable code note is because the earlier `check-type` against `valid-noise-value` means that t1 and t2 can never be 0.0d0.

appleby replied:
The "unreachable code" warning is because the type specifier `valid-noise-value` does not include 0.0d0. So the `check-types` above guarantee that `t1` and `t2` are non-zero.

If zero is invalid, you can just delete the `(if ..)` expression.

appleby replied:
whoops. should have refreshed the page. what zach said.

sophiaponte commented:
I went through the `src` files again to catch any formatting issues, and I think I've cleaned everything up at this point. The last thing that I need to do is squash my commits, and then I think this should be good to go unless anyone thinks otherwise :)
